substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/events/ExecutionSampleEvent.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.jfr.events;
+
+import java.util.concurrent.TimeUnit;
+
+import org.graalvm.nativeimage.CurrentIsolate;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.IsolateThread;
+import org.graalvm.nativeimage.StackValue;
+import org.graalvm.nativeimage.Threading;
+import org.graalvm.nativeimage.impl.ThreadingSupport;
+
+import com.oracle.svm.core.annotate.Uninterruptible;
+import com.oracle.svm.core.thread.JavaThreads;
+import com.oracle.svm.jfr.JfrEvents;
+import com.oracle.svm.jfr.JfrNativeEventWriter;
+import com.oracle.svm.jfr.JfrNativeEventWriterData;
+import com.oracle.svm.jfr.JfrNativeEventWriterDataAccess;
+import com.oracle.svm.jfr.JfrThreadState;
+import com.oracle.svm.jfr.JfrTicks;
+import com.oracle.svm.jfr.SubstrateJVM;
+
+public final class ExecutionSampleEvent {
+
+    private static long intervalMillis;
+    private static final ExecutionSampleEventCallback CALLBACK = new ExecutionSampleEventCallback();
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", calleeMustBe = false)
+    public static void tryToRegisterExecutionSampleEventCallback() {
+        if (SubstrateJVM.get().isEnabled(JfrEvents.ExecutionSample) && intervalMillis > 0) {
+            ImageSingletons.lookup(ThreadingSupport.class).registerRecurringCallback(intervalMillis, TimeUnit.MILLISECONDS, CALLBACK);
+        }
+    }
+
+    public static void setSamplingInterval(long type, long intervalMillis) {
+        if (type != JfrEvents.ExecutionSample.getId()) {
+            // JFR is currently only supporting ExecutionSample event, but this method is called
+            // during JFR startup, so we can't throw an error.
+            return;
+        }
+        if (intervalMillis < 0) {
+            throw new IllegalArgumentException("Sampling interval (" + intervalMillis + ") is not in valid range!");
+        }
+
+        ExecutionSampleEvent.intervalMillis = intervalMillis;
+    }
+
+    @Uninterruptible(reason = "Accesses a JFR buffer.")
+    public static void writeExecutionSample(IsolateThread isolateThread, Thread.State threadState) {
+        SubstrateJVM svm = SubstrateJVM.get();
+        if (SubstrateJVM.isRecording() && svm.isEnabled(JfrEvents.ExecutionSample)) {
+            JfrNativeEventWriterData data = StackValue.get(JfrNativeEventWriterData.class);
+            JfrNativeEventWriterDataAccess.initializeThreadLocalNativeBuffer(data);
+
+            JfrNativeEventWriter.beginEventWrite(data, false);
+            JfrNativeEventWriter.putLong(data, JfrEvents.ExecutionSample.getId());
+            JfrNativeEventWriter.putLong(data, JfrTicks.elapsedTicks());
+            JfrNativeEventWriter.putThread(data, isolateThread);
+            if (svm.isStackTraceEnabled(JfrEvents.ExecutionSample.getId())) {
```

Comment:
I think this pattern will be needed more frequently. It might be worth to create a method for it, e.g., `SubstrateJVM.getStackTraceId(long eventTypeId, int skipCount)`

Reply:
This will be useful, done!